src/webgpu/api/validation/resource_usages/textureUsageInRender.spec.ts

```diff
@@ -232,3 +237,133 @@ g.test('shader_stages_and_visibility')
       encoder.finish();
     });
   });
+
+g.test('replaced_binding')
+  .params(poptions('bindingType', kTextureBindingTypes))
+  .fn(async t => {
+    const { bindingType } = t.params;
+    const info = kTextureBindingTypeInfo[bindingType];
+    const bindingTexFormat = info.resource === 'storageTex' ? 'rgba8unorm' : undefined;
+
+    const view0 = t.createTexture().createView();
+    const view1 = t
+      .createTexture({ usage: GPUTextureUsage.STORAGE | GPUTextureUsage.SAMPLED })
+      .createView();
+
+    const bglEntries0: GPUBindGroupLayoutEntry[] = [
+      { binding: 0, visibility: GPUShaderStage.FRAGMENT, type: 'sampled-texture' },
+      {
+        binding: 1,
+        visibility: GPUShaderStage.FRAGMENT,
+        type: bindingType,
+        storageTextureFormat: bindingTexFormat,
+      },
+    ];
+    const bgEntries0: GPUBindGroupEntry[] = [
+      { binding: 0, resource: view0 },
+      { binding: 1, resource: view1 },
+    ];
+    const bindGroup0 = t.device.createBindGroup({
+      entries: bgEntries0,
+      layout: t.device.createBindGroupLayout({ entries: bglEntries0 }),
+    });
+
+    const bindGroup1 = t.device.createBindGroup({
+      entries: [{ binding: 0, resource: view1 }],
+      layout: t.device.createBindGroupLayout({
+        entries: [{ binding: 0, visibility: GPUShaderStage.FRAGMENT, type: 'sampled-texture' }],
+      }),
+    });
+
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: t.createTexture().createView(),
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+    pass.setBindGroup(0, bindGroup0);
+    pass.setBindGroup(0, bindGroup1);
+    pass.endPass();
+
+    // We should track the texture usages in bindings which are replaced by another setBindGroupi()
+    // call site upon the same index in the same render pass.
+    const success = bindingType === 'writeonly-storage-texture' ? false : true;
+    t.expectValidationError(() => {
+      encoder.finish();
+    }, !success);
+  });
+
+g.test('binding_in_bundle')
+  .params(
+    params()
+      .combine(poptions('typeInBundle', kTextureBindingTypes))
```

Comment:
Maybe this could be an additional parametrization of the `readwrite_upon_subresources` test that for each of the views says whether it is used in a bundle or not?

Reply:
readwrite_upon_subresources will be change in another pull request. I am considering that keep this test (a check for resource usage between bundle and render pass) but add a similar parameter as you suggested for "bundle" (pure bundle, no resource usage in render. The only exception is render-target's attachment), "render", "compute". Then most tests in this file will be expanded to 3 times as it is now. I am not sure whether that expansion are too many though. WDYT?

Reply:
Don't we need to test that bundles properly merge their subresource tracking with the main pass (including conflicts between two different bundles)? Doing it via another parametrization of `readwrite_upon_subresources` adds a x4 multiplier but gives us great coverage of that feature.

Reply:
Yeah. we can add that and get a lot more coverage. BTW, bundle is only introduced in render pass, so it should add a x3 multiplier: render pass with bundle, render pass without bundle, compute pass as I said. But you mentioned x4 multiplier. What's the 4th multiplier? compute pass with bundle doesn't exist per my understanding. Or you think we need to track one subresource in render pass, another in bundle for almost all test and make it be the 4th? (It is what this test is doing, but this test doesn't cover a lot combination, considering that we have already had a lot of combination somewhere. I just make sure resource in bundle is being tracked).

Reply:
I think the four are:

setBindGroup, executeBundles
setBindGroup, setBindGroup
executeBundles, executeBundles
executeBundles, setBindGroup

Reply:
Although like @Richard-Yunchao said the `executeBundles,  setBindGroup` and `setBindGroup, executeBundles` are similar so they could be deduplicated for only a x3 multiplier.

Reply:
One of the typeInPass options is render-target, which can't be inside a setBindGroup or an executeBundles. Unless we want to generalize it heavily I don't think this is the right test to test conflicts between bindgroups and bundles.

Reply:
That's ok if typeInPass is render-target for a bundle texture we can skip the test. But we definitely need to test interactions between a texture being used as an attachment and sampled in a bundle in that render pass.

Reply:
I revisit the test carefully and I personally think that it is not necessary to add "bindingInBundleOrNot" parameter to all tests in this file, because factors like subresources' combination or aspects or binding types are somehow irrelevant to bundle or not. And your suggestion seems to expand x3 or x4 in this test only. So I expanded this test to x3: 
setBindGroup in pass directly, executeBundle,
executeBundle, setBindGroup in pass directly,
executeBundle, excueteBundle. 
I didn't add non-bundle combination (setBindGroup in pass directly + setBindGroup in pass directly) because it was covered in many other tests.

Reply:
Sorry for the really slow reviews, I'm having a hard time staying on top of my inbox these days. It is important to test bundles in combination right now we have very imperfect coverage of [PassResourceUsageTracker::AddTextureUsage](https://source.chromium.org/chromium/chromium/src/+/master:third_party/dawn/src/dawn_native/PassResourceUsageTracker.cpp;drc=1562cab3f1eda927938f8f4a5a91991fefde66d3;l=67).

Usage tracking is a place where implementations are likely to do fancy optimizations and we need to provide them with great coverage. This is why I'm strongly suggesting that we do:

 - combination testing of render-pass render-pass interaction, render-bundle render-pass interaction in both direction, and render-bundle interactions between different render bundles
 - more varied shapes of subresources intersections for both resources

Reply:
Corentin, sounds like you want to expand the tests to 3 times for many other test cases (say `subresources_and_binding_types_combinations_for_color`) in this file, just as I suggested before. I thought about this carefully but right now I think it might be unnecessary. For example, subresource types combination and overlap measuring (different mip/array/plane slices, etc) is orthogonal to binding types combination (read-read, read-write, write-write usages). Because the implementation about whether a bunch of subresources overlap with another bunch of subresources is independent from the implementation about whether a binding type conflicts with another binding type. So I don't think the further combination of subresource types combination and binding types combinations make a lot of sense. So when we test subresource types/overlap, we should exhaust the subresource types/overlaps combination, with a few typical binding types combination. We don't need to iterate and exhaust every single binding types combinations for subresource types/overlap tests. It is the same when we test binding type combinations. But we did combination of these two orthogonal factors, which definitely improved the coverage, even though it might not make a lot of sense. 

`Bindings in bundle or not` is another factor, which is orthogonal to subresource types (They are set in BindGroup, we don't know they will be in bundle or not at that time) and binding types (they are set in BGL, we don't know they will be in bundle or not either), per my understanding. So I don't think that it makes a lot of sense to combine `bindings in bundle or not` with other factors (we can introduce a few typical cases of other factors, though). But it we exhaust the combination, it will definitely improves the coverage blindly. 

And we have other factors like visibility, sync scope (per draw/dispatch, or per pass, or out of pass), unused bindings, etc. Ideally we can combine all these factors and exhaust every possible combinations and make the tests cover every possible case, with the cost of huge complexity and terrible readability. So It's a balance, even though CTS is meant to cover every single corner (but it might not necessary to combine orthogonal/irrelevant factors). 

Well, if you think it is necessary, I will still try to combine `bindings in bundle or not` as a new parameter and expand other tests and see the complexity and readability after I added `compute` as a new parameter for tests to cover validation in compute.

Reply:
This comment is not yet addressed.

Reply:
Sorry for late response, Corentin. It is addressed at https://github.com/gpuweb/cts/pull/281. PTAL.